### ISSUE TITLE
Remove uses of `#[allow(dead_code)` in favor of names starting with  `_`

### ISSUE THREAD

**Issue by alamb** [apache/datafusion#13278]
TLDR is that we have some code (such as referenced below) that uses `#[allow(dead_code)]`  to permit a field that is only present so its `Drop` (destructor) can run. 

              Apparently this can be also written without disabling a lint:

```suggestion
    _reservation: MemoryReservation,
```

- `#[allow(dead_code)]` has disadvantage that it can be retained in the code even if the field becomes used
- `_reservation` has disadvantage that it doesn't look pretty

_Originally posted by @findepi in https://github.com/apache/datafusion/pull/13227#discussion_r1826521890_
            
@drauschenbach  notes: https://github.com/apache/datafusion/pull/13227#discussion_r1826582516

> Pretty or not, `_reservation` is the language-appropriate idiom.

Thus it would be great to update the code to use `_reservation` in the case above, and potentially look in the codebase for other similar uses of `#[allow(dead_code)]` and use the `_` prefix

**Comment by ding-young** [comment]
take
